### ISSUE TITLE
17.09 Crash muc_unsubscribe

### ISSUE THREAD

**Issue by Digithurst-RL** [processone/ejabberd#2270]
> What version of ejabberd are you using?

17.09

> What operating system (version) are you using?

Debian 9.3

> How did you install ejabberd (source, package, distribution)?

Distribution packages.

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behavior? What was the expected result?

ejabberd reports an error and crash log when trying to unsubscribe a user from a MUC-Sub enabled room. Subscription of the user works. The user is unsubscribed by an admin.
```xml
<iq
 type="set"
 id="1EA09B3A-AB16-4AC0-8A3A-C27D083FC050"
 from="secmess.ios.test.14df-100@dev.secretserver.de"
 to="30461@conference.dev.secretserver.de">
 <unsubscribe
  xmlns="urn:xmpp:mucsub:0"
  jid="secmess.ios.test.14df-102@dev.secretserver.de"/>
 </iq>
```

This is just an example of the message, not the exact IQ that resulted in the error, so names might not match 100%. It's the correct format though.
```
    =ERROR REPORT====
    ** State machine <0.855.1> terminating
    ** Last event in was {route,<<>>,{iq,<<"E4D478F9-D1BE-4DBD-8535-C957A48099B7">>,set,<<"de">>,{jid,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<"delete member test">>,<<"secmess.ios.test.14df-100%40telepax.d
    e">>,<<"dev.secretserver.de">>,<<"delete member test">>},{jid,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>},[{xmlel,<<"unsubscribe">>,[{<<"xmlns">>,<<"urn:xmpp:mucsu
    b:0">>},{<<"jid">>,<<"secmess.ios.test.14df-101@dev.secretserver.de">>}],[]}],#{ip => {0,0,0,0,0,65535,44048,2155}}}}
    ** When State == normal_state
    **      Data  == {state,<<"30461">>,<<"conference.dev.secretserver.de">>,<<"dev.secretserver.de">>,{muc,muc_create,muc_admin,muc_create},{jid,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>,<<"30461">>,<<"conference.dev.he
    althdataspace.de">>,<<>>},{config,<<>>,<<>>,true,true,true,anyone,true,true,false,true,true,true,false,true,false,false,true,false,<<>>,true,[moderator,participant,visitor],true,1800,200,false,<<>>,{0,nil},true},{dict,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{dict,3,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[[{<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>}|{subscriber,{jid,<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>},<<"secmess.ios.test.14df-101">>,[<<"urn:xmpp:mucsub:nodes:messages">>,<<"urn:xmpp:mucsub:nodes:presence">>]}]],[],[],[[{<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>}|{subscriber,{jid,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>},<<"secmess.ios.test.14df-100">>,[<<"urn:xmpp:mucsub:nodes:messages">>,<<"urn:xmpp:mucsub:nodes:presence">>]}]],[],[],[],[],[],[],[],[],[],[[{<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>}|{subscriber,{jid,<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>},<<"secmess.ios.test.14df-102">>,[<<"urn:xmpp:mucsub:nodes:messages">>,<<"urn:xmpp:mucsub:nodes:presence">>]}]],[],[]}}},{dict,3,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[[<<"secmess.ios.test.14df-102">>,{<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>}]],[[<<"secmess.ios.test.14df-101">>,{<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>}]],[[<<"secmess.ios.test.14df-100">>,{<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>}]],[],[],[],[],[],[],[],[],[],[],[],[]}}},nil,{dict,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{dict,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{dict,2,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[[{<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>}|{owner,<<>>}]],[],[],[],[],[],[],[],[],[],[[{<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>}|{owner,<<>>}]],[],[]}}},{lqueue,{{[],[]},0,unlimited},20},<<>>,<<>>,false,nil,none,undefined}
    ** Reason for termination =
    ** {function_clause,[{mod_muc_room,process_iq_mucsub,
[{jid,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<"delete member test">>,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<"delete member test">>},
{iq,<<"E4D478F9-D1BE-4DBD-8535-C957A48099B7">>,set,<<"de">>,{jid,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<"delete member test">>,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<"delete member test">>},{jid,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>},
  [{muc_unsubscribe,<<>>,{jid,<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>}}],
  #{ip => {0,0,0,0,0,65535,44048,2155}}},
{state,<<"30461">>,<<"conference.dev.secretserver.de">>,<<"dev.secretserver.de">>,{muc,muc_create,muc_admin,muc_create},{jid,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>},{config,<<>>,<<>>,true,true,true,anyone,true,true,false,true,true,true,false,true,false,false,true,false,<<>>,true,[moderator,participant,visitor],true,1800,200,false,<<>>,{0,nil},true},{dict,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{dict,3,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[[{<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>}|{subscriber,{jid,<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>},<<"secmess.ios.test.14df-101">>,[<<"urn:xmpp:mucsub:nodes:messages">>,<<"urn:xmpp:mucsub:nodes:presence">>]}]],[],[],[[{<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>}|{subscriber,{jid,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>},<<"secmess.ios.test.14df-100">>,[<<"urn:xmpp:mucsub:nodes:messages">>,<<"urn:xmpp:mucsub:nodes:presence">>]}]],[],[],[],[],[],[],[],[],[],[[{<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>}|{subscriber,{jid,<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>},<<"secmess.ios.test.14df-102">>,[<<"urn:xmpp:mucsub:nodes:messages">>,<<"urn:xmpp:mucsub:nodes:presence">>]}]],[],[]}}},{dict,3,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[[<<"secmess.ios.test.14df-102">>,{<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>}]],[[<<"secmess.ios.test.14df-101">>,{<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>}]],[[<<"secmess.ios.test.14df-100">>,{<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>}]],[],[],[],[],[],[],[],[],[],[],[],[]}}},nil,{dict,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{dict,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{dict,2,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[[{<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>}|{owner,<<>>}]],[],[],[],[],[],[],[],[],[],[[{<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>}|{owner,<<>>}]],[],[]}}},{lqueue,{{[],[]},0,unlimited},20},<<>>,<<>>,false,nil,none,undefined}],[{file,"src/mod_muc_room.erl"},{line,3705}]},{mod_muc_room,normal_state,2,[{file,"src/mod_muc_room.erl"},{line,298}]},{p1_fsm,handle_msg,10,[{file,"src/p1_fsm.erl"},{line,582}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}

    =CRASH REPORT====
      crasher:
        initial call: mod_muc_room:init/1
        pid: <0.855.1>
        registered_name: []
        exception exit: {{function_clause,[{mod_muc_room,process_iq_mucsub,[{jid,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<"delete member test">>,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<"delete member test">>},{iq,<<"E4D478F9-D1BE-4DBD-8535-C957A48099B7">>,set,<<"de">>,{jid,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<"delete member test">>,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<"delete member test">>},{jid,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>},[{muc_unsubscribe,<<>>,{jid,<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>}}],#{ip => {0,0,0,0,0,65535,44048,2155}}},{state,<<"30461">>,<<"conference.dev.secretserver.de">>,<<"dev.secretserver.de">>,{muc,muc_create,muc_admin,muc_create},{jid,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>,<<"30461">>,<<"conference.dev.secretserver.de">>,<<>>},{config,<<>>,<<>>,true,true,true,anyone,true,true,false,true,true,true,false,true,false,false,true,false,<<>>,true,[moderator,participant,visitor],true,1800,200,false,<<>>,{0,nil},true},{dict,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{dict,3,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[[{<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>}|{subscriber,{jid,<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>},<<"secmess.ios.test.14df-101">>,[<<"urn:xmpp:mucsub:nodes:messages">>,<<"urn:xmpp:mucsub:nodes:presence">>]}]],[],[],[[{<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>}|{subscriber,{jid,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>},<<"secmess.ios.test.14df-100">>,[<<"urn:xmpp:mucsub:nodes:messages">>,<<"urn:xmpp:mucsub:nodes:presence">>]}]],[],[],[],[],[],[],[],[],[],[[{<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>}|{subscriber,{jid,<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>,<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>},<<"secmess.ios.test.14df-102">>,[<<"urn:xmpp:mucsub:nodes:messages">>,<<"urn:xmpp:mucsub:nodes:presence">>]}]],[],[]}}},{dict,3,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[[<<"secmess.ios.test.14df-102">>,{<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>}]],[[<<"secmess.ios.test.14df-101">>,{<<"secmess.ios.test.14df-101">>,<<"dev.secretserver.de">>,<<>>}]],[[<<"secmess.ios.test.14df-100">>,{<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>}]],[],[],[],[],[],[],[],[],[],[],[],[]}}},nil,{dict,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{dict,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{dict,2,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[[{<<"secmess.ios.test.14df-100">>,<<"dev.secretserver.de">>,<<>>}|{owner,<<>>}]],[],[],[],[],[],[],[],[],[],[[{<<"secmess.ios.test.14df-102">>,<<"dev.secretserver.de">>,<<>>}|{owner,<<>>}]],[],[]}}},{lqueue,{{[],[]},0,unlimited},20},<<>>,<<>>,false,nil,none,undefined}],[{file,"src/mod_muc_room.erl"},{line,3705}]},{mod_muc_room,normal_state,2,[{file,"src/mod_muc_room.erl"},{line,298}]},{p1_fsm,handle_msg,10,[{file,"src/p1_fsm.erl"},{line,582}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]},[{p1_fsm,terminate,8,[{file,"src/p1_fsm.erl"},{line,760}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{17.09247}]}]}
        ancestors: ['mod_muc_dev.secretserver.de',ejabberd_gen_mod_sup,ejabberd_sup,<0.65.0>]
        message_queue_len: 0
        messages: []
        links: []
        dictionary: [{rand_seed,{#{bits => 58,jump => #Fun<rand.8.15449617>,next => #Fun<rand.5.15449617>,type => exrop,uniform => #Fun<rand.6.15449617>,uniform_n => #Fun<rand.7.15449617>,weak_low_bits => 1},[141315133450399713|54007253782846250]}},{'$internal_queue_len',0}]
        trap_exit: true
        status: running
        heap_size: 2586
        stack_size: 27
        reductions: 77392
      neighbours:
```

**Comment by badlop** [comment]
The code that allows a room admin to unsubscribe a subscriber was added in 17.06, so it should be present in your 17.09:
https://github.com/processone/ejabberd/commit/177d5fec862bb6b88c0d52b2592c75647a0eb954

But according to the error you see, that is not supported. Could it be you have two versions installed, and are running the old one? Can you confirm you are using 17.09? Can you see in the source code of that version, if the code mentioned in that commit is included?

**Comment by reitzig** [comment]
`ejabberdctl` shows version 17.09.

**Comment by badlop** [comment]
First I had tested with recent 18.01, and I didn't see the problem. Now I've compiled and installed 17.09 over the newer version, and the problem you mention appears. Then I remove all the xmpp- library versions, reinstall, and the problem disappears. ejabberd 17.09 wants to use xmpp-1.1.15. There were some changes in that library, in later releases, so mixing old ejabberd with newer xmpp library may produce that crash. Maybe you have other versions of that library installed?
